### PR TITLE
Use real subscription context

### PR DESCRIPTION
**Pull request by mitar**
Pull request for #159.

### CODE REVIEW

**Comment by mitar** [comment]
Because I use official code to publish, not accounts-base returning `null` instead of calling `ready()` is showing a warning. I opened a pull request to fix that: https://github.com/meteor/meteor/pull/6363

**Comment by mitar** [comment]
So this seems to work for me. I made it so that the publish context is really the official one. Also much of the official code is reused. I also fixed some small bugs which I discovered (for example, passing values as reference instead of copying it, and not removing fields in `changed` callback when value is `undefined`).

I would guess we could go even further and even make our "session" object the official one, and make it so that it keeps internal state of documents and all merging across subscriptions could be reused, we only make it so that when session sends a message, it does not really send a message to the client, but to our copy of client data on the server. So we override into the sesion's `sendAdded` and similar methods. I think this would make everything much easier and completely compatible with Meteor (with all diffing and merging and stuff done correctly).

**Comment by arunoda** [comment]
I think this is great. Merge box related stuff goes to a different PR I guess after this.
There's one more we need to do.

There are many apps (and yes accounts package) returns null thinking it'll stop the publication. We need to handle it with by overriding [`_publishHandlerResult`](https://github.com/meteor/meteor/blob/devel/packages/ddp-server/livedata_server.js#L1044).

On there, we need to convert `null` into `[]`.

**Comment by mitar** [comment]
> There are many apps (and yes accounts package) returns null thinking it'll stop the publication.

But how does Meteor handle this? How does it know that client-side does not have to wait for `.ready()` and it can render things?

**Comment by mitar** [comment]
BTW, `publishHandlerResult` is a separate function only in development branch. This was a pull request I made. So now it is easy to fix this, but it will not work for older versions of Meteor where `publishHandlerResult` is part of the `runHandler`.

**Comment by arunoda** [comment]
@mitar Okay got it. (yeah, the code looks fresh to me).
Then we may need to override handlers, but that's hacky.

In the accounts case, it's a universal publication and it has no meaning of `ready`. Some users do the similar in universal pubs and some don't do `ready` checks.

This is a pretty big issue if done in the FR related code. (Since this leads to 500ms delay in the server).

**Comment by mitar** [comment]
OK, I will address this by wrapping the handler itself.

**Comment by arunoda** [comment]
And then, you may need to wrap, `Meteor.publish` for runtime and publication loaded after Fast-Render.

File: lib/server/publish_context.js [hunk 1]
```diff
@@ -1,95 +1,81 @@
...
-  // Subscription class since they are private variables
-};
-
-PublishContext.prototype._addCursor = function(cursor) {
+PublishContext = function PublishContext(context, handler, subscriptionId, params, name) {
   var self = this;
-  cursor.rewind();
-  var collectionName = 
-    (cursor._cursorDescription)? cursor._cursorDescription.collectionName: null || //for meteor-collections
-    (cursor._collection)? cursor._collection._name: null; //for smart-collections
```
Comment (mitar) [inline lib/server/publish_context.js]: BTW, this is also something which was removed in the process. I do not know if this is problematic.

Comment (arunoda) [inline lib/server/publish_context.js]: That's fine. I assume there's no app use `smart-collections` now :D

**Comment by mitar** [comment]
I looked into this a bit more. In fact, universal subscriptions do not even have ready and stopped state. So this is in  fact a completely arbitrary choice for us when we decide it is "ready".

I would propose that instead of checking for `null`, we just assume all universal subscriptions are ready immediately after their handler exits.

**Comment by arunoda** [comment]
@mitar Okay. That's a good idea.
For other apps, we'll add some note here: https://github.com/kadirahq/fast-render/blob/master/lib/server/context.js#L102

**Comment by mitar** [comment]
Fixed.

File: lib/server/context.js [hunk 2]
```diff
@@ -108,26 +89,33 @@ Context.prototype.processPublication = function(publishHandler, publishContext,
     future.wait();
   }
 
+  //stop any runaway subscription
+  //this can happen if a publish handler never calls ready or stop, for example
+  //it does not hurt to call it multiple times
+  publishContext.stop();
```
Comment (mitar) [inline lib/server/context.js]: I also added this now.

File: lib/server/context.js [hunk 3]
```diff
@@ -99,7 +77,10 @@ Context.prototype.processPublication = function(publishHandler, publishContext,
     Meteor.setTimeout(function() {
       if (!future.isResolved()) {
         //publish handler failed to send ready signal in time
-        console.warn('Publish handler for', publishContext._subscription, 'sent no ready signal');
+        //maybe your non-universal publish handler is not calling this.ready()?
+        //or maybe it is returning null to signal empty publish?
+        //it should still call this.ready() or return an empty array []
```
Comment (arunoda) [inline lib/server/context.js]: I mean, we could something to the warn. Then user know something fishy is happening and how to start debug it.

Comment (mitar) [inline lib/server/context.js]: Feel free to modify it in any way you prefer. I like to keep logs short, but it is your package. :-)

Comment (arunoda) [inline lib/server/context.js]: Sure.

**Comment by arunoda** [comment]
Awesome.
This looks great to me. I'll take this in.

File: lib/server/context.js (hunk 3 above)
Comment (mitar) [inline lib/server/context.js]: Alternatively, you could display a longer warning only the first time.

**Comment by arunoda** [comment]
Merged and released as `v2.13.0`

**Pull request merged**